### PR TITLE
Fix verbatim control flow side effects

### PR DESCRIPTION
**Pull request by rodiazet**
- Reasoning: https://github.com/ethereum/solidity/pull/15535#discussion_r1811557042

### CODE REVIEW

File: libyul/backends/evm/EVMDialect.cpp [hunk 1]
```diff
@@ -547,6 +547,12 @@ BuiltinFunctionForEVM EVMDialect::createVerbatimFunction(size_t _arguments, size
 			);
 		}
 	);
+	// Worse control flow side effects because verbatim can do anything.
```
Comment (matheusaaguiar) [inline libyul/backends/evm/EVMDialect.cpp]:
```suggestion
	// Worst control flow side effects because verbatim can do anything.
```

**Review by matheusaaguiar — COMMENTED** [review]
Just a nitpick. And also, @cameel mentioned a changelog entry, but I am not sure about the actual bug he referred to.

File: libyul/backends/evm/EVMDialect.cpp (hunk 1 above)
Comment (rodiazet) [reply libyul/backends/evm/EVMDialect.cpp]: fixed.

**Comment by rodiazet** [comment]
> Just a nitpick. And also, @cameel mentioned a changelog entry, but I am not sure about the actual bug he referred to.

Also not sure. Typo fix pushed.

**Comment by cameel** [comment]
#15535 is merged so now we can pass these via the parameter we added there.

Might also be a good idea to define a helper called `worst()`, similarly to the one we already have for `SideEffects`.

**Review by cameel — CHANGES_REQUESTED** [review]
We discussed this bug on Wednesday and we do think that this may cause actual problems in the optimizer. E.g. UnusedStoreEliminator might mistakenly remove a store operation assuming that the code will revert, while a verbatim block actually makes it terminate successfully.

We need a repro test case to confirm it. I think that something like [`store_before_conditionally_terminating_function_call_revert.yul`](https://github.com/ethereum/solidity/blob/develop/test/libyul/yulOptimizerTests/unusedStoreEliminator/store_before_conditionally_terminating_function_call_revert.yul) should do it if you replace the `conditionallyStop()` call with a verbatim block that uses `RETURN`.

Still, we decided that we'll not treat it as an important bug (i.e. no blog post and no bug list entry, just a normal changelog entry), because doing something like this actually violates the prescribed usage rules for `verbatim`:

> - Control-flow should not jump into or out of verbatim blocks, but it can jump within the same verbatim block.

Instead, we should add a warning in [`verbatim` docs](https://docs.soliditylang.org/en/develop/yul.html#verbatim) saying that until 0.8.28 `verbatim` had this flaw. We could also clarify the point above to specifically say that reverting and returning within `verbatim` are not allowed.

**Comment by rodiazet** [comment]
> We discussed this bug on Wednesday and we do think that this may cause actual problems in the optimizer. E.g. UnusedStoreEliminator might mistakenly remove a store operation assuming that the code will revert, while a verbatim block actually makes it terminate successfully.
> 
> We need a repro test case to confirm it. I think that something like [`store_before_conditionally_terminating_function_call_revert.yul`](https://github.com/ethereum/solidity/blob/develop/test/libyul/yulOptimizerTests/unusedStoreEliminator/store_before_conditionally_terminating_function_call_revert.yul) should do it if you replace the `conditionallyStop()` call with a verbatim block that uses `RETURN`.
> 
> Still, we decided that we'll not treat it as an important bug (i.e. no blog post and no bug list entry, just a normal changelog entry), because doing something like this actually violates the prescribed usage rules for `verbatim`:
> 
> > * Control-flow should not jump into or out of verbatim blocks, but it can jump within the same verbatim block.
> 
> Instead, we should add a warning in [`verbatim` docs](https://docs.soliditylang.org/en/develop/yul.html#verbatim) saying that until 0.8.28 `verbatim` had this flaw. We could also clarify the point above to specifically say that reverting and returning within `verbatim` are not allowed.

I added such a test but even without this fix it passes. After debugging of `UnusedStoreEliminator` I guessing that it works because in `UnusedStoreEliminator::applyOperation` the store is considered as read from by the verbatim after it.

```
{
    let x := 0
    let y := 1
    sstore(x, y)
    verbatim_0i_0o(hex"600080F3")
    revert(0,0)
}
```

**Comment by cameel** [comment]
Nice. I looked at the implementation and I think you're right. So we're safe here. In that case we do not need the changelog entry or the warning.

I'd still add the test case though, even if it does not fail. Best to have this corner case covered.

And we should till clarify the docs to say that reverting and returning within verbatim are not allowed.

**Comment by rodiazet** [comment]
> Nice. I looked at the implementation and I think you're right. So we're safe here. In that case we do not need the changelog entry or the warning.
> 
> I'd still add the test case though, even if it does not fail. Best to have this corner case covered.
> 
> And we should till clarify the docs to say that reverting and returning within verbatim are not allowed.

Test added. Doc edited.

File: docs/yul.rst [hunk 2]
```diff
@@ -1095,6 +1095,7 @@ undefined behavior.
 
 - Control-flow should not jump into or out of verbatim blocks,
   but it can jump within the same verbatim block.
+- Reverting and returning within verbatim are *not* allowed.
```
Comment (cameel) [inline docs/yul.rst]: I'd merge it with the previous point. It's really a consequence of it.
```suggestion
  In particular, reverting or returning from the block is *not* allowed.
```

File: test/libyul/yulOptimizerTests/unusedStoreEliminator/store_before_terminating_verbatim_revert.yul [hunk 3]
```diff
@@ -0,0 +1,18 @@
+{
+    let x := 0
+    let y := 1
+    sstore(x, y)
+    verbatim_0i_0o(hex"600080F3")
+    revert(0,0)
```
Comment (cameel) [inline test/libyul/yulOptimizerTests/unusedStoreEliminator/store_before_terminating_verbatim_revert.yul:8]: Let's add some comments to make sure it's clear what the test is about:

```suggestion
    // opcodes: PUSH1 0x00 DUP1 RETURN
    // SSTORE is not redundant due to the RETURN hidden in the verbatim block. It must not be removed.
    verbatim_0i_0o(hex"600080F3")
    revert(0, 0)
```

**Review by cameel — DISMISSED** [review]
Looks good, but I have two small suggestions.

File: docs/yul.rst [hunk 4]
```diff
@@ -1095,6 +1095,7 @@ undefined behavior.
 
 - Control-flow should not jump into or out of verbatim blocks,
   but it can jump within the same verbatim block.
+- In particular, reverting or returning from the block is *not* allowed.
```
Comment (cameel) [inline docs/yul.rst]: Note that in https://github.com/ethereum/solidity/pull/15542#discussion_r1852823712 this was supposed to be merged with the bullet point above.

```suggestion
  In particular, reverting or returning from the block is *not* allowed.
```

**Review by cameel — DISMISSED** [review]

**Review by cameel — APPROVED** [review]

**Pull request merged**